cpp/src/arrow/util/ubsan.h

```diff
@@ -0,0 +1,47 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// Contains utilities for making UBSan happy.
+
+#pragma once
+
+#include <memory>
+
+namespace arrow {
+namespace util {
+
+/// \brief Returns maybe_null if not null or a non-null pointer to an arbitrary memory
+/// that shouldn't be dereferenced.
+///
+/// Memset/Memcpy are undefinfed when a nullptr is passed as an argument use this utility
+/// method to wrap locations where this could happen.
+///
+/// Note: Flatbuffers has UBSan warnings if a zero length vector is passed.
+/// https://github.com/google/flatbuffers/pull/5355 is trying to resolve them.
+template <typename T>
+inline T* MakeNonNull(T* maybe_null) {
+  if (maybe_null != nullptr) {
+    return maybe_null;
+  }
+
+  // Note: Stack based empty arrays are not allowed in c/c++.
+  static std::unique_ptr<uint8_t[]> non_null_filler(new uint8_t[0]);
```

Comment:
My C++ is weak here. But it seems like we might want just make a static variable (non-array) outside the header.

Reply:
Just use a non-empty array:
```
static char uint8_t empty[] = {0};
```

By the way, a function-static unique_ptr will involve synchronization to make its initialization thread-safe...

Reply:
just used a single char (did it outside of the method, but can move it inside if preferred).